### DISCUSSION TITLE
The best dashboards are built declarative with Markdown and SQL

### DISCUSSION THREAD

**Discussion by rinie** [observablehq/framework#1484] [General]
I was asked to keep this viewpoint separate from issues
https://github.com/observablehq/framework/issues/1346
https://github.com/observablehq/framework/issues/1405

Also observing the rise of DuckDb and Motherduck convince me that
'The best dashboards are built with code.' does not scale.
SQL/DuckDB versus Panda's or R enables colleagues and users that are fine using Excel adjusting 90% of a dashboard themselves.

For the remaining 10% I can help optimizing their SQL or use some trick like Hive Partitioning.
Promoting the 'any dataloader' path, reduces Observable Framework to a tool for developers.

DuckDB videos underline removing the barriers of entry that a Spark Cluster or even a Postgress DB has.
Just as Duck strives to 'just work' for CSV or Parquet instead of telling the users that their input is not valid.
IMHO you should target the modern browser stack + DuckDB WASM or as an implicit dataloader for CSV or SQL/Parquet.

Python/R etc can have their uses but just like Quarto it seems more valuable to 'just works with you less then perfect data in CSV, Excel or Parquet' and incremental data updates then 'you have to transform your data or write a Python script to start'

I can always optimize the SQL to perform better, non developers cannot improve a Python script...

**Comment by jaanaltosaar** [comment]
> I was asked to keep this viewpoint separate from issues #1346 #1405
> 
> Also observing the rise of DuckDb and Motherduck convince me that 'The best dashboards are built with code.' does not scale. 
> 
> I can always optimize the SQL to perform better, non developers cannot improve a Python script...

Strongly disagree with this - @onefact we have taught high school students, college students, postgraduate (Master's and PhD) and faculty in non-technical subjects (e.g. humanities) to improve python scripts. 

(And to write SQL; the homework from the first week of class last summer - colab [here](https://colab.research.google.com/github/onefact/datathinking.org-codespace/blob/main/notebooks/princeton-university/week-1-visualizing-33-million-phone-calls-in-new-york-city.ipynb) - was featured in the @motherduckdb blog: https://motherduck.com/blog/introducing-column-explorer/ and is currently used in the onboarding user flow as an example dataset.)

Please let me know if you need further examples - if we are wrong about this assumption or our experience we need to revise our data thinking curricula for the fall semester!

**Reply by rinie** [reply]
Sorry but my experience with only using YAML is that mere mortals don't grasp significant whitespace.
Which I also hate about python.
DuckDB https://duckdb.org/2022/05/04/friendlier-sql.html
DuckDB founders not specifically DBMS and Spark but I think YAML en Python have the same issues:
![image](https://github.com/observablehq/framework/assets/175597/d8a19e94-a73e-4a17-8140-e2722fb9e85d)

https://www.youtube.com/watch?v=9OFzOvV-to4&ab_channel=PositPBC

**Reply by jaanaltosaar** [reply]
I agree! 

> Sorry but my experience with only using YAML is that mere mortals don't grasp significant whitespace. Which I also hate about python. DuckDB https://duckdb.org/2022/05/04/friendlier-sql.html DuckDB founders not specifically DBMS and Spark but I think YAML en Python have the same issues: ![image](https://private-user-images.githubusercontent.com/175597/343176696-d8a19e94-a73e-4a17-8140-e2722fb9e85d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0ODM3NjMsIm5iZiI6MTcxOTQ4MzQ2MywicGF0aCI6Ii8xNzU1OTcvMzQzMTc2Njk2LWQ4YTE5ZTk0LWE3M2UtNGExNy04MTQwLWUyNzIyZmI5ZTg1ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyN1QxMDE3NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYzY2OTU4ODNkZTFhZGJjODkyMDQ2MzE2OTU1MDFjMDRkYjEzNWUzNDdkNjIyNDRjM2RiZWMyZjFhYTEyMTE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.A6cIe2yDDHH81NU_SG5a8a0n5ZXlHkZa1ahnp0axhh0)
> 
> https://www.youtube.com/watch?v=9OFzOvV-to4&ab_channel=PositPBC

😂😂 love that diagram of the hamburger! 

I also hate YAML with a passion, I think there are better tools like pkl from @apple and https://github.com/rails/jbuilder and python data classes that we have used to replace most YAML in our stack @onefact.

I think dbt has also really helped us move away from YAML, because it let us version control hospital prices from thousands of hospitals via duckdb SQL queries. 

Always happy to chat / discuss - and let me know if you’ve found better YAML replacements :) ideally statically type checked and hinted!
